tests: drivers: gpio: skip unsupported tests

cfriedt opened the pull request:
Some drivers may not support simultaneous input-output configuration or disconnect.

In such cases, the driver should return `-ENOTSUP` and the test should be skipped.

Fixes #46917

* Also update the suite to use the new ztest API
* Also correct error codes for related functions in `gpio.h`

Commit:
tests: drivers: gpio: direction: use new ztest api

Adjust suite to use new ztest API.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
tests: drivers: gpio: skip unsupported tests / drivers

Some drivers may not support simultaneous input-output
configuration or disconnect.

In such cases, the driver should return `-ENOTSUP` and the
test should be skipped.

Fixes #46917

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

yperess commented on tests/drivers/gpio/gpio_get_direction/src/main.c:
```diff
@@ -17,6 +17,11 @@ static void test_disconnect(void)
 	zassert_not_null(port, "device " TEST_DEV " not found");
 
 	rv = gpio_pin_configure(port, pin, flags);
+	if (rv == -ENOTSUP) {
```
Oh man, this would be a great candidate for #46961

yperess commented on tests/drivers/gpio/gpio_get_direction/src/main.c:
```diff
@@ -6,7 +6,7 @@
 #include "test_gpio_get_direction.h"
 
-static void test_disconnect(void)
+ZTEST(gpio_get_direction, test_disconnect)
```
Missing: `ZTEST_SUITE(gpio_get_direction, ...)`

cfriedt replied:
Weird - i distinctly remember adding that. Maybe I forgot to push

cfriedt replied:
May have been lost because i reordered commits. Fixed!

yperess commented on tests/drivers/gpio/gpio_get_direction/src/main.c:
```diff
@@ -17,6 +17,11 @@ static void test_disconnect(void)
 	zassert_not_null(port, "device " TEST_DEV " not found");
```
```
port = device_get_binding(TEST_DEV);
zassert_not_null(port, "device " TEST_DEV " not found");
```

Seems to be recurring, may I suggest (you can use zassert if you don't want to wait for the assume API, but then you should do the assert in the before function instead so it doesn't blow up your main thread):

```
struct gpio_get_direction_fixture {
  const struct device *port;
};

static void *gpio_get_direction_setup(void)
{
  static struct gpio_get_direction_fixture fixture;

  fixture.port = device_get_binding(TEST_DEV);
  zassume_not_null(port, "device " TEST_DEV " not found");

  return &fixture;
}

// Create the test suite and add the setup function
ZTEST_SUITE(gpio_get_direction, NULL, gpio_get_direction_setup, NULL, NULL, NULL);

// The _F tells it to use the fixture provided by the setup function
ZTEST_F(gpio_get_direction, test_disconnect)
{
  ...
  // Use fixture->port here
}
```

cfriedt replied:
I'll change it when the zassume work is done. I can see that would be helpful here.

PerMac commented:
I tried this PR but the failure is still the same on nrf boards

cfriedt commented:
> I tried this PR but the failure is still the same on nrf boards

@PerMac 
nRF boards do not yet support get direction, so that is expected.

Would be great if Nordic could add that support.

PerMac commented:
I see, but then it means the test description in its yaml requires some extra filtering options, so boards without getting direction won't be picked up

cfriedt commented:
@PerMac - skipping test at runtime for now :P I think proper filtering would likely require adding a Kconfig until there is support everywhere (e.g. `CONFIG_GPIO_DRIVER_SUPPORTS_DIRECTION`) but that seems like it could be overkill.

cfriedt commented:
@yperess - need to re-approve due to force-push, sorry.

PerMac commented:
TBH I am ambivalent about this solution. We will spend resources on our testing farm for building and executing a test, which in the end results only in skips. I can also imagine, that there is no other reasonable way for filtration in this case. If you'd managed to find a filtration option I'd appreciate it :)

cfriedt commented:
> TBH I am ambivalent about this solution. We will spend resources on our testing farm for building and executing a test, which in the end results only in skips. I can also imagine, that there is no other reasonable way for filtration in this case. However, if you'd managed to find a filtration option I'd appreciate it :)

Ideally @anangl would be able to add support for `gpio_get_direction()` in #47030, and then the test would be useful on nRF as well.

henrikbrixandersen commented:
> @PerMac - Ideally @anangl would be able to add support for `gpio_get_direction()` along with the fix for #47030, and then the test would be useful on nRF as well.

Had we included this test in the basic GPIO API test, and just skipped it for drivers not implementing this feature, the overhead would have been smaller.

cfriedt commented:
> > @PerMac - Ideally @anangl would be able to add support for `gpio_get_direction()` along with the fix for #47030, and then the test would be useful on nRF as well.
> 
> Had we included this test in the basic GPIO API test, and just skipped it for drivers not implementing this feature, the overhead would have been smaller.

If (when?) all platforms implement `gpio_get_direction()`, then I think it would make sense to merge it to the basic api test, because at that point, it would be part of the basic api and not `[EXPERIMENTAL]`.

yperess reviewed (changes requested):
One small nit

cfriedt commented on tests/drivers/gpio/gpio_get_direction/src/main.c:
```diff
@@ -6,108 +6,110 @@
+
+	fixture.pin = TEST_PIN;
+	fixture.port = device_get_binding(TEST_DEV);
+	zassert_not_null(fixture.port, "device " TEST_DEV " not found");
```
Will change.

I guess that could work though if there were state for each test. A failed setup function could signal all tests to fail in O(1).

yperess commented on tests/drivers/gpio/gpio_get_direction/src/main.c:
```diff
@@ -6,108 +6,117 @@
+{
+	struct gpio_get_direction_fixture *fixture = (struct gpio_get_direction_fixture *)arg;
+
+	fixture->port = device_get_binding(TEST_DEV);
```
No need to make a change here, but generally speaking the call to `device_get_binding` can be made in the setup since you only need it once per suite, then the assert can be in the before since it can't go in the setup.

cfriedt replied:
Fixed!

yperess commented on tests/drivers/gpio/gpio_get_direction/src/main.c:
```diff
@@ -6,108 +6,117 @@
+	}
 
-	rv = gpio_pin_configure(port, pin, flags);
 	zassert_equal(0, rv, "gpio_pin_configure() failed: %d", rv);
```
nit: can use `zassert_ok` here

cfriedt replied:
Fixed!

yperess commented on tests/drivers/gpio/gpio_get_direction/src/main.c:
```diff
@@ -6,108 +6,117 @@
 
-	rv = gpio_pin_is_output(port, pin);
+	rv = gpio_pin_is_output(fixture->port, fixture->pin);
 	zassert_equal(false, rv, "gpio_pin_is_output() failed: %d", rv);
```
nit: can use `zassert_ok` here

yperess commented on tests/drivers/gpio/gpio_get_direction/src/main.c:
```diff
@@ -6,108 +6,117 @@
 
-	rv = gpio_pin_is_output(port, pin);
+	rv = gpio_pin_is_output(fixture->port, fixture->pin);
 	zassert_equal(false, rv, "gpio_pin_is_output() failed: %d", rv);
```
nit: can use `zassert_ok` here

anangl commented on tests/drivers/gpio/gpio_get_direction/src/main.c:
```diff
@@ -6,108 +6,117 @@
-	rv = gpio_pin_is_input(port, pin);
-	if (rv == -ENOSYS) {
+	rv = gpio_pin_is_input(fixture->port, fixture->pin);
+	if (rv == -ENOTSUP) {
```
```suggestion
	if (rv == -ENOSYS) {
```

It looks like the description of `gpio_pin_is_input` needs to be corrected. It says: https://github.com/zephyrproject-rtos/zephyr/blob/3efce4c93938ebdd7d5d9b3ffdcd2b096490cea5/include/zephyr/drivers/gpio.h#L800-L803 while in `z_impl_gpio_port_get_direction` we've got: https://github.com/zephyrproject-rtos/zephyr/blob/3efce4c93938ebdd7d5d9b3ffdcd2b096490cea5/include/zephyr/drivers/gpio.h#L784-L785
Same story with `gpio_pin_is_output`.

cfriedt replied:
Good catch - I'll fix that in my current PR.

Commit:
api: gpio: correct doxygen for return values in error cases

If a method is not supported by the driver, it should return
-ENOSYS, not -ENOTSUP.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

cfriedt commented:
@yperess - please re-ack wyham
